docs/data/csv/auto_detection.md

```diff
@@ -108,13 +108,31 @@ The type detection works by attempting to convert the values in each column to t
 | TIMESTAMP |
 | VARCHAR   |
 
+Note everything can be cast to `VARCHAR`. This type has the lowest priority – i.e., columns are converted to `VARCHAR` if they cannot be cast to anything else. In [`flights.csv`](/data/flights.csv) the `FlightDate` column will be cast to a `DATE`, while the other columns will be cast to `VARCHAR`.
+
 The set of candidate types that should be considered by the CSV reader can be explicitly specified using the [`auto_type_candidates`]({% link docs/data/csv/overview.md %}#auto_type_candidates-details) option. 
 
-Note everything can be cast to `VARCHAR`. This type has the lowest priority – i.e., columns are converted to `VARCHAR` if they cannot be cast to anything else. In [`flights.csv`](/data/flights.csv) the `FlightDate` column will be cast to a `DATE`, while the other columns will be cast to `VARCHAR`.
+In addition to the default set of candidate types, other types that may be specified using the `auto_type_candidates` options are:
+
+<div class="narrow_table monospace_table"></div>
+
+|   Types   |
+|-----------|
+| DECIMAL   |
+| FLOAT     |
+| INTEGER   |
+| SMALLINT  |
+| TINYINT   |
+
+Even though the set of data types that can be automatically detected may appear quite limited, the CSV reader can configured to read arbitrarily complex types by using the `types`-option described in the next section. 
+
+Type detection can be entirely disabled by using the `all_varchar` option. If this is set all columns will remain as `VARCHAR` (as they originally occur in the CSV file).
+
+#### Overriding Type Detection 
 
-The detected types can be individually overridden using the `types` option. This option takes either a list of types (e.g., `types = [INTEGER, VARCHAR, DATE]`) which overrides the types of the columns in-order of occurrence in the CSV file. Alternatively, `types` takes a `name` → `type` map which overrides options of individual columns (e.g., `types = {'quarter': INTEGER}`).
+The detected types can be individually overridden using the `types` option. This option takes either a list of type definitions (e.g., `types = ['INTEGER', 'VARCHAR', 'DATE']`) which overrides the types of the columns in-order of occurrence in the CSV file. Alternatively, `types` takes a `name` → `type` map which overrides options of individual columns (e.g., `types = {'quarter': 'INTEGER'}`).
 
-The type detection can be entirely disabled by using the `all_varchar` option. If this is set all columns will remain as `VARCHAR` (as they originally occur in the CSV file).
+The set of column types that may be specified using the `types` option is not as limited as the types available for the `auto_type_candidates` option: Any valid type definition is acceptable to the `types`-option. (To get a valid type definition, use the [`typeof()`]({% link docs/sql/functions/utility.md %}#typeofexpression) function, or use the `column_type` column  of the [`DESCRIBE`]({% link docs/guides/meta/describe.md %}) result.)  
```

Comment:
You can also use `select columns from sniff_csv('path/to/file.csv')` to get a struct with column names and types.

Reply:
Thanks!

Reply:
Hi Pedro! I checked that, and it looks to me like sniff_csv returns the same subset of types that is supported for the `types` option. They work of course, but what I was trying to convey with this piece of doc is that you can specify wildly exotic types like `STRUCT(a INTEGER[3], b VARCHAR[3])` and the CSV reader will actually swallow it and produce such values. 

I tested it like this:

```
copy (
  select *
  from   test_all_types()
)
to "C:\roland\projects\duckdb-bag-of-tricks\csv_dump_and_restore\test_all_types.csv";
```
```
set variable test_all_types_types = (
  select list(column_type)
  from   (
    describe 
    select *
    from   test_all_types()
  )
);
```

and then test whether all types are indeed reproduced:

```
select * from (describe from test_all_types())
except
select * from (describe from    read_csv(
  'C:\roland\projects\duckdb-bag-of-tricks\csv_dump_and_restore\test_all_types.csv'
  , types=getvariable("test_all_types_types")
  )
);
```
(results in 0 rows, as expected)

and then check if all values are read correctly:

```
select * exclude ("timestamp_tz","union")
from    read_csv(
          'C:\roland\projects\duckdb-bag-of-tricks\csv_dump_and_restore\test_all_types.csv'
        , types=getvariable("test_all_types_types")
        )
except
select * exclude ("timestamp_tz","union")
from test_all_types()
```

timestamp_tz and union columns do not produce the correct value but I decided to ignore that as that's probably some unrelated issue, and the principle still stands that the csv reader can reliably read arbitrary complex data types.

I was surprised by this actually and its hard to express how awesome that is! 

Anyway, I think your additional comment is also good but I'd add something to the effect that you then might lose type information whereas typeof and describe will give you a more complete definition of the type. 

Of course, your method of using the sniffer is completely appropriate when you only have CSV files, and no additional metadata. But I would then add some text that explains these two situations, because the feature of being to read arbitrary types from csv is really quite something!

Reply:
@pdet Is my assessment correct that (save for these exceptions of union and timestamp_tz in this particular example), the CSV reader can handle deserialization of arbitrily complex types, provided the right type definition is passed to the reader? 

If so then I can amend the docs to describe this along with some type on when to use the sniffer detected types instead.